scipy/signal/_peak_finding.py

```diff
@@ -233,6 +236,733 @@ def argrelextrema(data, comparator, axis=0, order=1, mode='clip'):
+    >>> plt.plot(x)
+    >>> plt.plot(peaks, x[peaks], "x")
+    >>> plt.vlines(peaks, ymin=contour_heights, ymax=x[peaks])
+    """
+    x = np.asarray(x)
+    peaks = np.asarray(peaks)
+    if peaks.size == 0:
+        # Handle empty peaks
+        return np.array([]), np.array([], dtype=int), np.array([], dtype=int)
+
+    if x.ndim != 1:
+        raise ValueError('`x` must have exactly one dimension')
+    if peaks.ndim != 1:
+        raise ValueError('`peaks` must have exactly one dimension')
+    if x.size <= peaks.max():
+        raise ValueError('an index in `peaks` exceeds the size of `x`')
+    if not np.issubdtype(peaks.dtype, np.integer):
+        raise ValueError('`peaks` must be an array of integers')
+    if wlen is not None and wlen < 3:
+            raise ValueError('`wlen` must be at least 3')
+
+    # Prepare return arguments
+    prominences = np.zeros(peaks.size)
+    left_bases = np.zeros(peaks.size, dtype=int)
+    right_bases = np.zeros(peaks.size, dtype=int)
+
+    for i, peak in enumerate(peaks):
+        # If wlen is twice the size of x the symmetric window always covers
+        # the full x
+        if wlen is not None and wlen < x.size * 2:
+            wlen = int(wlen)
+            # Calculate window borders around the evaluated peak
+            wleft = peak - wlen // 2
+            wright = peak + wlen // 2
+            # Handle border cases
+            wleft = 0 if wleft < 0 else wleft
+            wright = x.size if x.size < wright else wright
+            # Use slice for prominence calculation
+            window = x[wleft:wright]
+            # Correct peak position in x
+            peak -= wleft
+        else:
+            # Use full x for prominence calculation
+            window = x
+            wleft = 0
+
+        # Positions where window is larger than current peak height
+        greater_peak = np.where(window > window[peak])[0]
+
+        try:
+            # Nearest position to the left of peak with
+            # window[left] > window[peak]
+            left = greater_peak[greater_peak < peak].max()
+        except ValueError:
+            left = 0
+        try:
+            # Nearest position to right of peak with
+            # window[right] > window[peak]
+            right = greater_peak[greater_peak > peak].min()
+        except ValueError:
+            right = None
+
+        # Base indices to the left and right of peak in window
+        left_bases[i] = window[left:peak].argmin() + left
+        right_bases[i] = window[peak:right].argmin() + peak
+
+        # Calculate lowest contour and its vertical distance to peak
+        lowest_contour = max(window[left_bases[i]], window[right_bases[i]])
+        prominences[i] = window[peak] - lowest_contour
+
+        # Correct window offset
+        left_bases[i] += wleft
+        right_bases[i] += wleft
+
+    return prominences, left_bases, right_bases
+
+
+def peak_widths(x, peaks, rel_height=0.5, **kwargs):
```

Comment:
`**kwargs` is generally to be avoided unless there is a good reason to use it. For example, it is:

- less inspection-friendly than `prominences=None, left_bases=None, right_bases=None, wlen=None`
- less robust to user typos in `kwarg` entry like `peak_widths(x, peaks, rel_heigth=1.0)`

Reply:
```
def peak_widths(x, peaks, rel_height=0.5, wlen=None, prominence_out=None): 
    ...
```
Would a signature like this reasonable? `prominence_out` would expect tuple matching the output of `peak_prominences` which is currently provided through `**kwargs`. I find this still better than using all three keyword arguments directly because they are expected to be always passed together (only one of them would still require recalculation of the two others).
I'm not to happy with the name `prominence_out` though...

Reply:
Yes let's find a better name. `prominence_data`?